torch/distributed/fsdp/flatten_params_wrapper.py

```diff
@@ -74,13 +73,31 @@ def _pre_load_state_dict_hook(
             _replace_by_prefix(state_dict, k, prefix + last_part)
 
 
+class Interval(NamedTuple):
+    begin: int
+    end: int
+
+    def is_empty(self):
+        return not (self.begin >= 0 and self.end >= 0 and self.end >= self.begin)
+
+
+ParamOffset = Interval
+SharedParamInfo = Tuple[str, str, nn.Module, str, nn.Module, str]
+
+
 class ParamInfo(NamedTuple):
     module_name: str
     module: nn.Module
     param_name: str
 
 
-class ShardMetadata(NamedTuple):
+class ParamMetadata(NamedTuple):
+    names: List[str]
+    shapes: List[torch.Size]
+    numels: List[int]
```

Comment:
do we need to keep track of `numels`? Can we get it from shape just by multiplying, or is it more convenient to have this attr?

Reply:
We can infer numels from shapes. This is from the original code, and I'm not sure if meta is actually being used. Can do another refactor later.